File: selfdrive/proclogd/proclogd.cc
#include <dirent.h>
#include <sys/resource.h>
#include <sys/time.h>

#include <algorithm>
#include <cassert>
#include <climits>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <functional>
#include <memory>
#include <iostream>
#include <sstream>
#include <unordered_map>
#include <utility>

#include "cereal/messaging/messaging.h"
#include "selfdrive/common/timing.h"
#include "selfdrive/common/util.h"

ExitHandler do_exit;

struct CPUTime {
  int id;
  unsigned long utime, ntime, stime, itime;
  unsigned long iowtime, irqtime, sirqtime;
};

struct PidStat {
    int pid, ppid, processor;
    char state;
    unsigned long utime, stime;
    long cutime, cstime, priority, nice, num_threads;
    unsigned long long starttime;
    unsigned long vms, rss;
    std::string name;
};

struct ProcCache {
  int pid;
  std::string name;
  std::vector<std::string> cmdline;
  std::string exe;
};

const double jiffy = sysconf(_SC_CLK_TCK);
const size_t page_size = sysconf(_SC_PAGE_SIZE);

// parse /proc/stat
std::vector<CPUTime> parseProcStat(std::istream &stream) {
  std::vector<CPUTime> cpu_times;
  std::string line;
  while (std::getline(stream, line)) {
    if (util::starts_with(line, "cpu ")) {
      // cpu total
    } else if (util::starts_with(line, "cpu")) {
      // specific cpu
      CPUTime t = {};
      sscanf(line.data(), "cpu%d %lu %lu %lu %lu %lu %lu %lu",
             &t.id, &t.utime, &t.ntime, &t.stime, &t.itime, &t.iowtime, &t.irqtime, &t.sirqtime);
      cpu_times.push_back(t);
    } else {
      break;
    }
  }
  return cpu_times;
}

// parse /proc/meminfo
std::unordered_map<std::string, uint64_t> parseMemInfo(std::istream &stream) {
  std::unordered_map<std::string, uint64_t> mem_info;
  std::string line, key;
  while (std::getline(stream, line)) {
    std::istringstream l(line);
    uint64_t val = 0;
    l >> key >> val;
    if (!l.fail()) {
      mem_info[key] = val * 1024;
    }
  }
  return mem_info;
}


const ProcCache &getProcExtraInfo(int pid, const std::string &name) {
  static std::unordered_map<pid_t, ProcCache> proc_cache;
  ProcCache &cache = proc_cache[pid];
  if (cache.pid != pid || cache.name != name) {
    cache.pid = pid;
    cache.name = name;
    cache.exe = util::readlink(util::string_format("/proc/%d/exe", pid));

    // null-delimited cmdline arguments to vector
    std::string cmdline_s = util::read_file(util::string_format("/proc/%d/cmdline", pid));
    const char *cmdline_p = cmdline_s.c_str();
    const char *cmdline_ep = cmdline_p + cmdline_s.size();

    // strip trailing null bytes
    while ((cmdline_ep - 1) > cmdline_p && *(cmdline_ep - 1) == 0) {
      cmdline_ep--;
    }

    cache.cmdline.clear();
    while (cmdline_p < cmdline_ep) {
      std::string arg(cmdline_p);
      cache.cmdline.push_back(arg);
      cmdline_p += arg.size() + 1;
    }
  }
  return cache;
}

// parse /proc/pid/stat
std::optional<PidStat> parsePidStat(const std::string &stat) {
  PidStat p;
  char name[PATH_MAX] = {};
  int count = sscanf(stat.data(),
                     "%d (%1024[^)]) %c %d %*d %*d %*d %*d %*d %*d %*d %*d %*d "
                     "%lu %lu %ld %ld %ld %ld %ld %*d %lld "
                     "%lu %lu %*d %*d %*d %*d %*d %*d %*d "
                     "%*d %*d %*d %*d %*d %*d %*d %d",
                     &p.pid, name, &p.state, &p.ppid,
                     &p.utime, &p.stime, &p.cutime, &p.cstime, &p.priority, &p.nice, &p.num_threads, &p.starttime,
                     &p.vms, &p.rss, &p.processor);
  if (count == 15) {
    p.name = name;
    return p;
  }
  return std::nullopt;
}

namespace Test {

void test_parseProcStat() {
  {
    std::string stat = "cpu 5 1 2 3 4 5 6 7 8\ncpu0 0 1 2 3 4 5 6 7 8\ncpu1 0 1 2 3 4 5 6 7 8";
    std::istringstream stream(stat);
    auto stats = parseProcStat(stream);
    assert(stats.size() == 2);
    assert(stats[0].id == 0 && stats[0].utime == 0 && stats[0].ntime == 1);
    assert(stats[1].id == 1 && stats[0].utime == 0 && stats[0].ntime == 1);
  }
  {
    std::string stat = util::read_file("/proc/stat");
    std::istringstream stream(stat);
    auto stats = parseProcStat(stream);
    assert(stats.size() == sysconf(_SC_NPROCESSORS_ONLN));
    for (int i = 0; i < stats.size(); ++i) {
      assert(stats[i].id == i);
    }
  }
}

void test_parseMemInfo() {
  {
    std::string stat = "MemTotal:    1024 kb\nMemFree:    10 kb\n";
    std::istringstream stream(stat);
    auto stats = parseMemInfo(stream);
    assert(stats["MemTotal:"] = 1024 * 1024 && stats["MemFree:"] == 1024 * 10);
  }
  {
    std::string stat = "MemTotal:   kb \nMemFree:    10 kb\n";
    std::istringstream stream(stat);
    auto stats = parseMemInfo(stream);
    assert(stats.find("MemTotal:") == stats.end());
    assert(stats["MemFree:"] == 1024 * 10);
  }
  {
    std::string stat = util::read_file("/proc/meminfo");
    std::istringstream stream(stat);
    auto stats = parseMemInfo(stream);
    std::string keys[] = {"MemTotal:", "MemFree:", "MemAvailable:", "Buffers:", "Cached:", "Active:", "Inactive:", "Shmem:"};
    for (auto &key : keys) {
      assert(stats.find(key) != stats.end());
    }
  }
}

void test_parsePidStat() {
  std::string normal = "167151 (cat) R 7297 167151 7297 34818 167151 4194304 92 0 0 0 0 0 0 0 20 0 1 0 303166 11563008 130 18446744073709551615 93994557333504 93994557359153 140732482410736 0 0 0 0 0 0 0 0 0 17 20 0 0 0 0 0 93994557377168 93994557378752 93994559090688 140732482412879 140732482412899 140732482412899 140732482416619 0";
  auto stat = parsePidStat(normal);
  assert(stat && stat->name == "cat" && stat->pid == 167151);

  std::string with_space = "167151 (cat 123) R 7297 167151 7297 34818 167151 4194304 92 0 0 0 0 0 0 0 20 0 1 0 303166 11563008 130 18446744073709551615 93994557333504 93994557359153 140732482410736 0 0 0 0 0 0 0 0 0 17 20 0 0 0 0 0 93994557377168 93994557378752 93994559090688 140732482412879 140732482412899 140732482412899 140732482416619 0";
  stat = parsePidStat(with_space);
  assert(stat && stat->name == "cat 123" && stat->pid == 167151);

  std::string less = "167151 (cat 123) R 7297 167151";
  stat = parsePidStat(less);
  assert(!stat);

  std::string more = "167151 (cat 123) R 7297 167151 7297 34818 167151 4194304 92 0 0 0 0 0 0 0 20 0 1 0 303166 11563008 130 18446744073709551615 93994557333504 93994557359153 140732482410736 0 0 0 0 0 0 0 0 0 17 20 0 0 0 0 0 93994557377168 93994557378752 93994559090688 140732482412879 140732482412899 140732482412899 140732482416619 0 1 2 3";
  stat = parsePidStat(less);
  assert(!stat);
}

int test_Parse() {
  Test::test_parseProcStat();
  Test::test_parseMemInfo();
  Test::test_parsePidStat();
  std::cout << "test ok" << std::endl;
  return 0;
}

}  // namespace Test

void buildCPUTimes(cereal::ProcLog::Builder &builder) {
  std::vector<CPUTime> cpu_times;

  std::ifstream stream("/proc/stat");
  std::vector<CPUTime> stats = parseProcStat(stream);

  auto log_cpu_times = builder.initCpuTimes(cpu_times.size());
  for (int i = 0; i < cpu_times.size(); ++i) {
    auto l = log_cpu_times[i];
    auto r = cpu_times[i];
    l.setCpuNum(r.id);
    l.setUser(r.utime / jiffy);
    l.setNice(r.ntime / jiffy);
    l.setSystem(r.stime / jiffy);
    l.setIdle(r.itime / jiffy);
    l.setIowait(r.iowtime / jiffy);
    l.setIrq(r.irqtime / jiffy);
    l.setSoftirq(r.irqtime / jiffy);
  }
}

void buildMemoryInfo(cereal::ProcLog::Builder &builder) {
  std::ifstream smem("/proc/meminfo");
  auto mem_info = parseMemInfo(smem);

  auto mem = builder.initMem();
  mem.setTotal(mem_info["MemTotal:"]);
  mem.setFree(mem_info["MemFree:"]);
  mem.setAvailable(mem_info["MemAvailable:"]);
  mem.setBuffers(mem_info["Buffers:"]);
  mem.setCached(mem_info["Cached:"]);
  mem.setActive(mem_info["Active:"]);
  mem.setInactive(mem_info["Inactive:"]);
  mem.setShared(mem_info["Shmem:"]);
}

void buildProcesses(cereal::ProcLog::Builder &builder) {
  std::vector<PidStat> procs_info;
  struct dirent *de = NULL;
  DIR *d = opendir("/proc");
  assert(d);
  while ((de = readdir(d))) {
    if (isdigit(de->d_name[0])) {
      int pid = atoi(de->d_name);
      auto stat = parsePidStat(util::read_file(util::string_format("/proc/%d/stat", pid)));
      if (stat) {
        procs_info.push_back(*stat);
      }
    }
  }
  closedir(d);

  auto lprocs = builder.initProcs(procs_info.size());
  for (size_t i = 0; i < procs_info.size(); i++) {
    auto l = lprocs[i];
    PidStat &r = procs_info[i];
    l.setPid(r.pid);
    l.setState(r.state);
    l.setPpid(r.ppid);
    l.setCpuUser(r.utime / jiffy);
    l.setCpuSystem(r.stime / jiffy);
    l.setCpuChildrenUser(r.cutime / jiffy);
    l.setCpuChildrenSystem(r.cstime / jiffy);
    l.setPriority(r.priority);
    l.setNice(r.nice);
    l.setNumThreads(r.num_threads);
    l.setStartTime(r.starttime / jiffy);
    l.setMemVms(r.vms);
    l.setMemRss((uint64_t)r.rss * page_size);
    l.setProcessor(r.processor);
    l.setName(r.name);

    const ProcCache &extra_info = getProcExtraInfo(r.pid, r.name);
    l.setExe(extra_info.exe);

    auto lcmdline = l.initCmdline(extra_info.cmdline.size());
    for (size_t i = 0; i < lcmdline.size(); i++) {
      lcmdline.set(i, extra_info.cmdline[i]);
    }
  }
}

int main(int argc, char** argv) {
  if (argc == 2 && strcmp(argv[1], "--test") == 0) {
    return Test::test_Parse();
  }
  setpriority(PRIO_PROCESS, 0, -15);
  PubMaster publisher({"procLog"});

  while (!do_exit) {
    MessageBuilder msg;
    auto procLog = msg.initEvent().initProcLog();
    buildCPUTimes(procLog);
    buildMemoryInfo(procLog);
    buildProcesses(procLog);
    publisher.send("procLog", msg);

    util::sleep_for(2000);  // 2 secs
  }

  return 0;
}

File: system/proclogd.py
#!/usr/bin/env python3
import os
from typing import NoReturn, TypedDict

from cereal import messaging
from openpilot.common.realtime import Ratekeeper
from openpilot.common.swaglog import cloudlog

JIFFY = os.sysconf(os.sysconf_names['SC_CLK_TCK'])
PAGE_SIZE = os.sysconf(os.sysconf_names['SC_PAGE_SIZE'])


def _cpu_times() -> list[dict[str, float]]:
  cpu_times: list[dict[str, float]] = []
  try:
    with open('/proc/stat') as f:
      lines = f.readlines()[1:]
    for line in lines:
      if not line.startswith('cpu') or len(line) < 4 or not line[3].isdigit():
        break
      parts = line.split()
      cpu_times.append({
        'cpuNum': int(parts[0][3:]),
        'user': float(parts[1]) / JIFFY,
        'nice': float(parts[2]) / JIFFY,
        'system': float(parts[3]) / JIFFY,
        'idle': float(parts[4]) / JIFFY,
        'iowait': float(parts[5]) / JIFFY,
        'irq': float(parts[6]) / JIFFY,
        'softirq': float(parts[7]) / JIFFY,
      })
  except Exception:
    cloudlog.exception("failed to read /proc/stat")
  return cpu_times


def _mem_info() -> dict[str, int]:
  keys = ["MemTotal:", "MemFree:", "MemAvailable:", "Buffers:", "Cached:", "Active:", "Inactive:", "Shmem:"]
  info: dict[str, int] = dict.fromkeys(keys, 0)
  try:
    with open('/proc/meminfo') as f:
      for line in f:
        parts = line.split()
        if parts and parts[0] in info:
          info[parts[0]] = int(parts[1]) * 1024
  except Exception:
    cloudlog.exception("failed to read /proc/meminfo")
  return info


_STAT_POS = {
  'pid': 1,
  'state': 3,
  'ppid': 4,
  'utime': 14,
  'stime': 15,
  'cutime': 16,
  'cstime': 17,
  'priority': 18,
  'nice': 19,
  'num_threads': 20,
  'starttime': 22,
  'vsize': 23,
  'rss': 24,
  'processor': 39,
}

class ProcStat(TypedDict):
  name: str
  pid: int
  state: str
  ppid: int
  utime: int
  stime: int
  cutime: int
  cstime: int
  priority: int
  nice: int
  num_threads: int
  starttime: int
  vms: int
  rss: int
  processor: int


def _parse_proc_stat(stat: str) -> ProcStat | None:
  open_paren = stat.find('(')
  close_paren = stat.rfind(')')
  if open_paren == -1 or close_paren == -1 or open_paren > close_paren:
    return None
  name = stat[open_paren + 1:close_paren]
  stat = stat[:open_paren] + stat[open_paren:close_paren].replace(' ', '_') + stat[close_paren:]
  parts = stat.split()
  if len(parts) < 52:
    return None
  try:
    return {
      'name': name,
      'pid': int(parts[_STAT_POS['pid'] - 1]),
      'state': parts[_STAT_POS['state'] - 1][0],
      'ppid': int(parts[_STAT_POS['ppid'] - 1]),
      'utime': int(parts[_STAT_POS['utime'] - 1]),
      'stime': int(parts[_STAT_POS['stime'] - 1]),
      'cutime': int(parts[_STAT_POS['cutime'] - 1]),
      'cstime': int(parts[_STAT_POS['cstime'] - 1]),
      'priority': int(parts[_STAT_POS['priority'] - 1]),
      'nice': int(parts[_STAT_POS['nice'] - 1]),
      'num_threads': int(parts[_STAT_POS['num_threads'] - 1]),
      'starttime': int(parts[_STAT_POS['starttime'] - 1]),
      'vms': int(parts[_STAT_POS['vsize'] - 1]),
      'rss': int(parts[_STAT_POS['rss'] - 1]),
      'processor': int(parts[_STAT_POS['processor'] - 1]),
    }
  except Exception:
    cloudlog.exception("failed to parse /proc/<pid>/stat")
    return None

class ProcExtra(TypedDict):
  pid: int
  name: str
  exe: str
  cmdline: list[str]


_proc_cache: dict[int, ProcExtra] = {}


def _get_proc_extra(pid: int, name: str) -> ProcExtra:
  cache: ProcExtra | None = _proc_cache.get(pid)
  if cache is None or cache.get('name') != name:
    exe = ''
    cmdline: list[str] = []
    try:
      exe = os.readlink(f'/proc/{pid}/exe')
    except OSError:
      pass
    try:
      with open(f'/proc/{pid}/cmdline', 'rb') as f:
        cmdline = [c.decode('utf-8', errors='replace') for c in f.read().split(b'\0') if c]
    except OSError:
      pass
    cache = {'pid': pid, 'name': name, 'exe': exe, 'cmdline': cmdline}
    _proc_cache[pid] = cache
  return cache


def _procs() -> list[ProcStat]:
  stats: list[ProcStat] = []
  for pid_str in os.listdir('/proc'):
    if not pid_str.isdigit():
      continue
    try:
      with open(f'/proc/{pid_str}/stat') as f:
        stat = f.read()
      parsed = _parse_proc_stat(stat)
      if parsed is not None:
        stats.append(parsed)
    except OSError:
      continue
  return stats


def build_proc_log_message(msg) -> None:
  pl = msg.procLog

  procs = _procs()
  l = pl.init('procs', len(procs))
  for i, r in enumerate(procs):
    proc = l[i]
    proc.pid = r['pid']
    proc.state = ord(r['state'][0])
    proc.ppid = r['ppid']
    proc.cpuUser = r['utime'] / JIFFY
    proc.cpuSystem = r['stime'] / JIFFY
    proc.cpuChildrenUser = r['cutime'] / JIFFY
    proc.cpuChildrenSystem = r['cstime'] / JIFFY
    proc.priority = r['priority']
    proc.nice = r['nice']
    proc.numThreads = r['num_threads']
    proc.startTime = r['starttime'] / JIFFY
    proc.memVms = r['vms']
    proc.memRss = r['rss'] * PAGE_SIZE
    proc.processor = r['processor']
    proc.name = r['name']

    extra = _get_proc_extra(r['pid'], r['name'])
    proc.exe = extra['exe']
    cmdline = proc.init('cmdline', len(extra['cmdline']))
    for j, arg in enumerate(extra['cmdline']):
      cmdline[j] = arg

  cpu_times = _cpu_times()
  cpu_list = pl.init('cpuTimes', len(cpu_times))
  for i, ct in enumerate(cpu_times):
    cpu = cpu_list[i]
    cpu.cpuNum = ct['cpuNum']
    cpu.user = ct['user']
    cpu.nice = ct['nice']
    cpu.system = ct['system']
    cpu.idle = ct['idle']
    cpu.iowait = ct['iowait']
    cpu.irq = ct['irq']
    cpu.softirq = ct['softirq']

  mem_info = _mem_info()
  pl.mem.total = mem_info["MemTotal:"]
  pl.mem.free = mem_info["MemFree:"]
  pl.mem.available = mem_info["MemAvailable:"]
  pl.mem.buffers = mem_info["Buffers:"]
  pl.mem.cached = mem_info["Cached:"]
  pl.mem.active = mem_info["Active:"]
  pl.mem.inactive = mem_info["Inactive:"]
  pl.mem.shared = mem_info["Shmem:"]


def main() -> NoReturn:
  pm = messaging.PubMaster(['procLog'])
  rk = Ratekeeper(0.5)
  while True:
    msg = messaging.new_message('procLog', valid=True)
    build_proc_log_message(msg)
    pm.send('procLog', msg)
    rk.keep_time()


if __name__ == '__main__':
  main()
